src/v/storage/storage_resources.cc

```diff
@@ -91,9 +92,21 @@ void storage_resources::update_allowance(uint64_t total, uint64_t free) {
     _falloc_step = calc_falloc_step();
 }
 
+void storage_resources::update_min_checkpoint_bytes() {
+    if (_partition_count) {
+        // Limit the checkpoint frequency to 2x what it would be under
+        // uniform traffic to all partitions. This permits us to overshoot
+        // target_replay_bytes by approx 50% under non-uniform writes.
+        _min_checkpoint_bytes = _target_replay_bytes() / (_partition_count * 2);
```

Comment:
does this have a weird failure mode with small replay size bytes and large partition count?

Reply:
We should get a zero in that case, which is the same as having no lower bound: checkpoints will go ahead whenever the total replay bytes semaphore is exhausted